File: DataStructures.py
import time
import numpy as np
import tqdm

class DataStructures:
    def __init__(self, filepath, progressbar=False, timer=False):
        """ Creates the data structures from the given file path.

        Args:
            filepath (str): Path to the .fmi file to create the data structures from.
            progressbar (bool, optional): Whether to show a progress bar. This will hurt performance. Defaults to False.
            timer (bool, optional): Whether to show a timer. Defaults to False.
        """

        if timer:
            time_start = time.time()

        with open(filepath, "r") as file:
            for _ in range(5):
                next(file)

            amount_nodes = int(file.readline())
            amount_edges = int(file.readline())
            
            node_array = np.zeros((amount_nodes, 4), dtype=np.longdouble)
            edge_array = np.zeros((amount_edges, 3), dtype=np.int32)

            # Read the nodes and edges into the arrays.
            for i in tqdm.tqdm(range(amount_nodes), disable=not progressbar, desc="1/3 Reading nodes"):
                line = file.readline().split()
                node_array[i] = [int(line[0]), np.longdouble(line[2]), np.longdouble(line[3]), 0]

            for i in tqdm.tqdm(range(amount_edges), disable=not progressbar, desc="2/3 Reading edges"):
                line = file.readline().split()
                edge_array[i] = [line[0], line[1], line[2]]
            
            # Get the offset for each node as described in the specification to improve performance for Dijkstra's algorithm.
            for i in tqdm.tqdm(range(len(edge_array)), disable=not progressbar, desc="3/3 Calculating offsets"):
                if i == 0:
                    node_array[edge_array[i][0]][3] = 0
                else:
                    if edge_array[i][0] != edge_array[i - 1][0]:
                        node_array[edge_array[i][0]][3] = i
                
                    

        self.node_array = node_array
        self.sort_node_array_by_latitude()
        
        self.edge_array = edge_array
            
        if timer:
            time_end = time.time()
            print(f"Elapsed: {str(time_end - time_start)} seconds.")
            
            
    def sort_node_array_by_latitude(self):
        """ Sorts the node array by latitude.
        """

        sorted_indices = np.argsort(self.node_array[:, 1])
        self.node_array = self.node_array[sorted_indices]

    def find_nearest_lat_lon(self, target_lat, target_lon, tolerance=0.1):
        
        # TODO: Documentation needed.
        
        data = self.node_array
        # Create a numpy array with all coordinates within the latitude tolerance range.
        latitude_range = np.logical_and(data[:, 1] >= target_lat - tolerance, data[:, 1] <= target_lat + tolerance)
        filtered_data = data[latitude_range]

        if len(filtered_data) == 0:
            return None

        # Create a numpy array with all coordinates within the longitude tolerance range.
        longitude_range = np.logical_and(filtered_data[:, 2] >= target_lon - tolerance, filtered_data[:, 2] <= target_lon + tolerance)
        filtered_data = filtered_data[longitude_range]

        if len(filtered_data) == 0:
            return None

        # Calculate distances to all points within the square-like area.
        distances = np.sqrt((filtered_data[:, 1] - target_lat) ** 2 + (filtered_data[:, 2] - target_lon) ** 2)

        # Find the nearest point within the square-like area.
        nearest_index = np.argmin(distances)
        nearest_combination = filtered_data[nearest_index]

        return nearest_combination

            
    def get_node_array(self):
        """ Returns the node array.

        Returns:
            np.ndarray: Node array.
        """

        return self.node_array
    
    def get_edge_array(self):
        """ Returns the edge array.

        Returns:
            np.ndarray: Edge array.
        """

        return self.edge_array
            
if __name__ == "__main__":
    ds = DataStructures("toy.fmi")
    edges = ds.get_edge_array()
    nodes = ds.get_node_array()
    print(nodes[3][3])

File: Java/DataStructures.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import java.math.BigDecimal;

public class DataStructures {
    private int[][] nodesWithOffset;
    private BigDecimal[][] nodesOrderedByLatitude;
    private int[][] edges;

    public DataStructures(String filePath, boolean progressBar, boolean timer) {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            for (int i = 0; i < 5; i++) {
                    reader.readLine();
                }

            int amountNodes = Integer.parseInt(reader.readLine());
            int amountEdges = Integer.parseInt(reader.readLine());

            this.nodesWithOffset = new int[amountNodes][3];
            this.nodesOrderedByLatitude = new BigDecimal[amountNodes][3];
            this.edges = new int[amountEdges][3];

            // Read the nodes and edges into the arrays.
            for (int i = 0; i < amountNodes; i++) {
                String[] line = reader.readLine().split(" ");
                this.nodesOrderedByLatitude[i] = new BigDecimal[] {new BigDecimal(line[1]), new BigDecimal(line[2]), new BigDecimal(line[3])};
                this.nodesWithOffset[i][0] = Integer.parseInt(line[0]);
            }

            for (int i = 0; i < amountEdges; i++) {
                String[] line = reader.readLine().split(" ");
                this.edges[i] = new int[] {Integer.parseInt(line[0]), Integer.parseInt(line[1]), Integer.parseInt(line[2])};
            }

            // Get the offset for each node as described in the specification to improve performance for Dijkstra's algorithm as well as the amount of edges for each node.
            for (int i = 0; i < edges.length; i++) {
                if (i == 0) {
                    this.nodesWithOffset[this.edges[i][0]][1] = 0;
                    this.nodesWithOffset[this.edges[i][0]][2] = 1;
                } else {
                    if (this.edges[i][0] != this.edges[i - 1][0]) {
                        this.nodesWithOffset[this.edges[i - 1][0]][2] = i - this.nodesWithOffset[this.edges[i - 1][0]][1];
                        this.nodesWithOffset[this.edges[i][0]][1] = i;
                        this.nodesWithOffset[this.edges[i][0]][2] = 1;
                    } else {
                        this.nodesWithOffset[this.edges[i][0]][2] += 1;
                    }
                }
            }

            sortNodeArrayByLatitude();

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        
    private void sortNodeArrayByLatitude() {
        Arrays.sort(this.nodesOrderedByLatitude, (a, b) -> a[1].compareTo(b[1]));
    }

    public BigDecimal[] findNearestLatLon(BigDecimal targetLat, BigDecimal targetLon, BigDecimal tolerance) {
        BigDecimal[] nearestCombination = null;
        BigDecimal[][] data = this.nodesOrderedByLatitude;

        // Create a BigDecimal array with all coordinates within the latitude tolerance range.
        BigDecimal[][] filteredData = Arrays.stream(data)
            .filter(row -> row[1].subtract(targetLat).abs().compareTo(tolerance) <= 0)
            .toArray(BigDecimal[][]::new);

        if (filteredData.length == 0) {
            return nearestCombination;
        }

        // Create a BigDecimal array with all coordinates within the BigDecimalitude tolerance range.
        filteredData = Arrays.stream(filteredData)
            .filter(row -> row[2].subtract(targetLon).abs().compareTo(tolerance) <= 0)
            .toArray(BigDecimal[][]::new);

        if (filteredData.length == 0) {
            return nearestCombination;
        }

        // Calculate distances to all points within the square-like area.
        BigDecimal minDistance = BigDecimal.valueOf(Double.MAX_VALUE);

        for (BigDecimal[] row : filteredData) {
            BigDecimal distance = BigDecimal.valueOf(Math.sqrt(Math.pow(row[1].subtract(targetLat).doubleValue(), 2) + Math.pow(row[2].subtract(targetLon).doubleValue(), 2)));
            if (distance.compareTo(minDistance) < 0) {
                minDistance = distance;
                nearestCombination = row;
            }
        }

        return nearestCombination;
    }

    public BigDecimal[][] getNodesOrderedByLatitude() {
        return this.nodesOrderedByLatitude;
    }

    public int[][] getNodesWithOffset() {
        return this.nodesWithOffset;
    }

    public int[][] getEdges() {
        return this.edges;
    }

    public static void main(String[] args) {
        DataStructures ds = new DataStructures("stuttgart.fmi", false, false);
        int[][] edges = ds.getEdges();
        BigDecimal[][] nodesOrderedByLatitude = ds.getNodesOrderedByLatitude();
        int[][] nodesWithOffset = ds.getNodesWithOffset();

        /*for (BigDecimal[] node : nodesOrderedByLatitude) {
            System.out.println(Arrays.toString(node));
        }*/

        BigDecimal[] node = ds.findNearestLatLon(new BigDecimal(48.832639), new BigDecimal(9.223725), new BigDecimal(0.001));
        System.out.println(Arrays.toString(node));
    }
}